add errorline.jl

CMGreenspon opened the pull request:
Function for parsing inputs to easily make a ribbons (https://ggplot2.tidyverse.org/reference/geom_ribbon.html)  or errorbar https://www.mathworks.com/help/matlab/ref/errorbar.html) plot while allowing for easily controlling error type and also handles NaNs.

BeastyBlacksmith commented on .vscode/settings.json:
```diff
@@ -0,0 +1 @@
+{}
```
This shouldn't be here

BeastyBlacksmith commented on src/errorline.jl:
```diff
@@ -0,0 +1,179 @@
+        The second dimension is treated as the repeated observations and error is computed along this dimension. If the 
+        matrix has a 3rd dimension this is treated as a new group.
+
+    ErrorStyle (symbol - *:Ribbon* or :Classic) - determines whether to use a ribbon style or stick style error representation.
```
```suggestion
    error_style (symbol - *:ribbon* or :classic) - determines whether to use a ribbon style or stick style error representation.
```
In julia usually only types and modules have upper CamelCase

BeastyBlacksmith replied:
Same applies for the other keyword arguments

BeastyBlacksmith commented on src/errorline.jl:
```diff
@@ -0,0 +1,179 @@
+"""
+errorline
+
+function compute_error(y::AbstractMatrix, CenterType::Symbol, ErrorType::Symbol, Percentiles::Vector{Int64})
```
```suggestion
function compute_error(y::AbstractMatrix, CenterType::Symbol, ErrorType::Symbol, Percentiles::Vector{Int})
```
This currently fails the tests on 32bit machines. Any reason this couldn't be `percentiles::AbstractVector`?

BeastyBlacksmith commented on src/errorline.jl:
```diff
@@ -0,0 +1,179 @@
+        if !any(size(y) .== length(x))
+            error("Size of x and y do not match")
+        elseif ndims(y) == 2 && size(y,1) != length(x) && size(y,2) == length(x) # Check if y needs to be transposed or transmuted
+            y = y'
```
`y'` is the adjoint you might want to use `transpose` or `permutedims` here

BeastyBlacksmith commented on src/errorline.jl:
```diff
@@ -0,0 +1,179 @@
+    end
+
+    # Parse different color type
+    if typeof(GroupColor) == Symbol || typeof(GroupColor) == RGB{Float64}
```
```suggestion
    if GroupColor isa Symbol || GroupColor isa RGB{Float64}
```
What about `RGBA`?

BeastyBlacksmith reviewed (changes requested):
Thank you for your contribution. I got some stylistic and functional remarks.
Additionally we should think about whether this recipe needs to live here or could also be added to `Plots.jl` itself.
Currently, I don't see a reason why not.

CMGreenspon commented:
I believe I've addressed all the changes. The error you got was because NaNMath only accepts Floats so I've added a conversion check. I also added the final distribution mode.

BeastyBlacksmith reviewed (approved):
So far this looks fine to me, there should be at least one example in the README as Documentation. Would be good to have another pair of eyes here.

CMGreenspon commented:
I'll add an example or two to the readme in an hour or so

On Tue, Aug 23, 2022, 7:17 AM Simon Christ ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> So far this looks fine to me, there should be at least one example in the
> README as Documentation. Would be good to have another pari of eyes here.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaPlots/StatsPlots.jl/pull/513#pullrequestreview-1082054902>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF6PLUSKKXDEWKY43SF6VI3V2S6L5ANCNFSM57IEPJGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

t-bltg commented on src/errorline.jl:
```diff
@@ -0,0 +1,236 @@
+        given then it will use a different color for each group.
+
+    secondarycolor (Symbol, RGB, :matched - *:Gray60*) - When using stick mode this will allow for the setting of the stick color.
+        If ":matched" is given then the color of the sticks with match that of the main line.
```
```suggestion
        If `:matched` is given then the color of the sticks with match that of the main line.
```

Use backquotes for references to keywords.

t-bltg commented on src/errorline.jl:
```diff
@@ -0,0 +1,236 @@
+        the default colorscheme. If one value is given then it will use that color for all groups. If multiple colors are
+        given then it will use a different color for each group.
+
+    secondarycolor (Symbol, RGB, :matched - *:Gray60*) - When using stick mode this will allow for the setting of the stick color.
```
```suggestion
    secondarycolor (`Symbol`, `RGB`, `:matched` - *:Gray60*) - When using stick mode this will allow for the setting of the stick color.
```

t-bltg commented on src/errorline.jl:
```diff
@@ -0,0 +1,236 @@
+        The second dimension is treated as the repeated observations and error is computed along this dimension. If the
+        matrix has a 3rd dimension this is treated as a new group.
+
+    error_style (symbol - *:ribbon*, :stick, :spaghetti) - determines whether to use a ribbon style or stick style error
```
```suggestion
    error_style (`Symbol` - *:ribbon*, :stick, :spaghetti) - determines whether to use a ribbon style or stick style error
```

t-bltg commented on src/errorline.jl:
```diff
@@ -0,0 +1,236 @@
+            num_obs = size(y,2)
+            if num_obs > numsecondarylines
+                sub_sample_idx = sample(1:num_obs, numsecondarylines, replace=false)
+                y_sub_sample = y[:,sub_sample_idx,g]
```
I don't know what convention is used in `StatsPlots` (maybe we need an autoformatter here as well), but I find `y[:, sub_sample_idx, g]` more readable.

t-bltg reviewed (commented):
I'm not a `StatsPlots` user so I'm only going to add stylistic comments here. Otherwise, I don't see why it cannot go in, since this PR is tested and documented.

CMGreenspon commented on src/errorline.jl:
```diff
@@ -0,0 +1,236 @@
+            end
+
+        elseif errorstyle == :stick
+            x_offset = (extrema(x)[2] - extrema(x)[1]) * stickwidth
```
Thank you for this! I come from matlab and googling range in julia didn't give me what I wanted to this is perfect!

CMGreenspon commented on src/errorline.jl:
```diff
@@ -0,0 +1,236 @@
+        y = float(y)
+    end
+    # First compute the center
+    if centertype == :mean
+        y_central =  mapslices(NaNMath.mean, y, dims=2)
+    elseif centertype == :median
+        y_central =  mapslices(NaNMath.median, y, dims=2)
+    end
```
I don't love that line 60 is a single line, I think it makes it awkward to read, but if broken into two lines I think that's pretty good so I'll do that unless that's against convention?

CMGreenspon commented:
![plot_36](https://user-images.githubusercontent.com/24966610/186193984-8601301e-8e03-48e3-98c3-159cf40c9e13.svg)
Example plot

CMGreenspon commented:
![ExamplePlot](https://user-images.githubusercontent.com/24966610/186654791-e8e2bab8-b31c-4ee5-ae80-3c0ba7cf8c62.svg)
Rename spaghetti to plume

CMGreenspon commented:
![ExamplePlot](https://user-images.githubusercontent.com/24966610/186655231-2b7b9e37-0beb-4796-ad08-cbb84020ffd8.svg)
